### PR TITLE
Shorten projection types in inlay hints

### PR DESCRIPTION
**Pull request by WaffleLapkin**
In examples like this, where we have a projection type, which is the only associated item with its name in a trait that is in scope:
```rust
trait Trait {
    type Assoc;

    fn f(self) -> Self::Assoc;
}

fn f<T: Trait>(x: T) {
    let y /*here*/ = x.f();
    _ = y;
}
```
Show `T::Assoc` instead of `<T as Trait>::Assoc`.

Rationale: making inlay hints less verbose is generally good, `Trait` is not that useful there.

-----

I'm not sure if the implementation is good and correct, but it seems to be working at least 😅

### CODE REVIEW

File: crates/hir-ty/src/display.rs [hunk 1]
```diff
@@ -323,10 +343,43 @@ impl HirDisplay for ProjectionTy {
             return write!(f, "{}", TYPE_HINT_TRUNCATION);
         }
 
+        let assoc_name =
+            f.db.type_alias_data(from_assoc_type_id(self.associated_ty_id)).name.to_string();
```
Comment (lowr) [inline crates/hir-ty/src/display.rs:347]: `Name` implements `Display` and `PartialEq`, so can we avoid string allocation here (and L.367)?

**Review by lowr — COMMENTED** [review]
While I understand the motivation and the logic looks correct, I'm a bit concerned about the performance here; iterating through every associated item of every trait in scope sounds a bit expensive especially for large files. I guess it's okay for typical codebases as it's fairly rare to show projection types, but we might want to assess the effect or add a configuration to opt it out.

Another thing is that it might be somewhat confusing unless the self type of the projection is either placeholder type or ADT. `T::Assoc` and `Vec<i32>::Assoc` seem nice, but `(T, U)::Assoc`, `impl Trait + Clone::Assoc`, and `&dyn Trait + Send::Assoc` look foreign to me. Also, `<&T as Trait>::Assoc` would be displayed as `&T::Assoc`, which is indistinguishable from `&<T as Trait>::Assoc`.

File: crates/hir-ty/src/display.rs (hunk 1 above)
Comment (WaffleLapkin) [reply crates/hir-ty/src/display.rs]: Yep. Somehow missed the fact that implements `PartialEq`, oops.

**Comment by Veykril** [comment]
I agree with lowr's comment here, this does seem like a performance pitfall and the ambiguity cases mentioned aren't too nice either. So I am inclined to close this for now.

**Comment by WaffleLapkin** [comment]
Okay, I don't have enough time/energy to work on this, but hopefully I'll eventually try to revive this :)

**Pull request closed without merging**